### PR TITLE
feat(datastore) Add start and stop, and stop starting on configure and clear

### PR DESCRIPTION
**Pull request by richardmcclellan**
Summary
 - Added `void DataStore::start(Action onComplete, Consumer<DataStoreException> onError)` API 
 - Added `void DataStore::stop(Action onComplete, Consumer<DataStoreException> onError)` API
 - Added `Completable RxDataStore::start()` and `Completable RxDataStore::stop()` API
 - Modified `DataStore::configure()` so that it does not call start anymore
 - Modified `DataStore::clear()` so that it does not call start anymore
 - The following scenarios now bubble up an exception to the user provided `Consumer<DataStoreException>`, instead of just logging a warning:
   - [Timeout acquiring lock in Orchestrator::start()](https://github.com/aws-amplify/amplify-android/blob/89ed53d3f2b887c6b23439c776be133fa1ac0819/aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java#L181)
   - [Timeout starting DataStore](https://github.com/aws-amplify/amplify-android/blob/c612d56181ba4a686ff2da4fc8a8913050f9863f/aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java#L501)
   - [Any other unexpected error while starting DataStore](https://github.com/aws-amplify/amplify-android/blob/c612d56181ba4a686ff2da4fc8a8913050f9863f/aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java#L503)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:128]:
```suggestion
        // Configure DataStore with an empty config (All defaults)
```

File: aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:228]: Line 223 still says "synchronization processes are restarted"; could you update the Javadoc?

File: aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java:165]: This won't actually fail the test. jUnit will only report a test failure if the assertion error is raised on the test runner's thread. This will crash the `new Thread`, but it won't get reported.

If you _don't_ rely on the `fail` for the test to be useful, I'd suggest changing this `fail` to a comment, like:
```java
// This doesn't matter, we'll catch the failure later when we validate (thing you'll validate later).
```

If you _are_ expecting this `fail` to catch a bad state, you'd have to coalesce it back into control flow on the test runner thread, somehow.

File: core/src/main/java/com/amplifyframework/datastore/DataStoreCategory.java
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/datastore/DataStoreCategory.java:195]: Maybe we should order these nearer to `configure(...)` and `initialize(...)` in the document, since they are more lifecycle-oriented?

File: core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java:231]: Suggest a few tweaks to language to:
1. Remove use of "it", which increases cognitive buffering in dense technical writing, and
2. Talk explicitly about synchronization with a remote system. Currently, there is not enough differentiation from `configure()` or `initialize()`. 
```suggestion
     * Starts the DataStore's synchronization with a remote system, if DataStore is configured to support
     * remote synchronization. This only needs to be called if you wish to start the synchronization eagerly.
     * If you don't call start(), the synchronization will start automatically, prior to executing any other
     * operations (#query, #save, #delete, #observe).
```

I might suggest going back and updating `configure()`, `initialize()` (well, it's on the category), and `clear()`, to more explicitly highlight their relationship to the new `start()` and `stop()`.

File: core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java:237]: Similar thoughts as above, about the `onComplete`, here. But also: what's the DX for this callback, why would someone use it? Shouldn't `start(...)` wait until model synchronization is fully started, to invoke this? Otherwise, why wouldn't I just always use the `READY` event? And if I should always prefer `READY`, then couldn't we remove this callback? Lastly, we now have:
1. Configured
2. Initialized
3. Started
4. Ready

All of which mean different things. That might be okay _internally_ to our code, but I wonder if we can eliminate ~2 of these entirely from the user's view.

File: testutils/src/main/java/com/amplifyframework/testutils/VoidResult.java
Comment (jamesonwilliams) [inline testutils/src/main/java/com/amplifyframework/testutils/VoidResult.java:22]: Nice!! I think there are a few places where I could have used this, around the code base. I can go back and patch them up, now that you've added this.

**Review by jamesonwilliams — COMMENTED** [review]
Code changes look very clean, nice @richardmcclellan! Thoughts below are mostly about how we communicate these changes in the Javadoc.

File: aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java
Comment (richardmcclellan) [reply aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:228]: The comment was actually wrong before - the previous implementation restarted synchronization immediately).   

The new behavior matches the comment now.  `clear` won't restart synchronization, but the next interaction with DataStore will.

File: core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java
Comment (rjuliano) [reply core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java:239]: I've struggled with this one. If the intent is to align with JS (as much as possible), `start()` really means `startSyncEngine()`, but the naming is besides the point here. One of the things we discussed is that for the mobile platforms, blocking execution during the entirety of the initialization of sync engine could lead to poor UX.

In the past, I've advocated for blocking only until `StorageObserver` starts since that's the bare minimum we would need to ensure no customer data is lost. That being said, that would give the onComplete callback yet another meaning, so I'm not too sure about that.

Looking at the plugin code, it looks like the `onComplete` callback basically lets them know that we've received their request to start the sync engine. I commented under the start method of the plugin as well, but there may be a scenario which both onComplete and onError can fire for the same call, which could lead to confusion.

Maybe the "easy" answer is to have no callbacks on start and stop, and tell devs to listen to sync engine events on the Hub.

**Review by rjuliano — COMMENTED** [review]
This looks good. Left a couple of comments for your consideration

File: core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java
Comment (richardmcclellan) [reply core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java:239]: I've been scratching my head for a while about this as well :).

Here's how `start()` is implemented in this PR:
 - `onComplete` is called after:
   - plugin initialization completes (if needed)
   - `storageObserver.startObservingStorageChanges` completes.
- `onError` is called:
  - if initialization doesn't complete within 2 seconds
  - if orchestrator start stop lock is not acquired within 2 seconds
  - if `storageObserver.startObservingStorageChanges` doesn't complete within 2 seconds

Realistically, if `onError` is called, the error is not going to be something clients can handle anyway.  It'll probably only be called because of a bug in the library.

A basic app built with DataStore should look something like this (`start()` is not even a critical component):

```
private void showListOfTodos() {
    runQuery(); // Show locally cached data (or local synced data if already synced)
    Amplify.Hub.subscribe(HubChannel.DATASTORE,
            hubEvent -> DataStoreChannelEventName.READY.toString().equals(hubEvent.getName()),
            hubEvent -> runQuery()); // Show synced data
}

private void runQuery() {
    Amplify.DataStore.query(Todo.class,
            result -> { /* Display results on UI */},
            error -> { }
    );
}
```

Customers only need to call `start()` if they want to pre-cache data before displaying it to the end user.  They would call start(), presumably on app launch, and then call `query` whenever the end user navigates to a view that requires data.   Based on this, I don't think onComplete / onError make much sense, so I will remove them.  

For `stop()`, there are two customer use cases:
1. Close subscription sockets to save battery when DataStore is not needed, such as on app close.
2. Force a re-evaluation of sync query predicates. A customer would immediately call `query(...)` after `stop()`. Since we have the startStopSemaphore locking the stop operation, this should work just fine.

Callbacks aren't particularly useful for either of those use cases.   I'll bring this up with iOS/JS in the DataStore sync meeting tomorrow, and if we all agree, we will remove the callbacks.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java
Comment (richardmcclellan) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:265]: > Could both onComplete and onError fire for the same call? I think onComplete will always fire, but if an exception is thrown by orchestrator.start() the onError(...) callback will also be invoked which could lead to devs assuming that if onComplete() fires, they're OK to move on with the rest of their call flow.

Hmm, did I mess this up?   I think `onComplete` will _not_ be called on error, because there's a `return` on line 263, right?

Comment (rjuliano) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:265]: Nevermind. For whatever reason I thought the onComplete.call() was outside of the curly braces.

File: core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java
Comment (richardmcclellan) [reply core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java:239]: I spoke with the iOS and JS teams about this a few days ago, and they opposed removing these callbacks. 

The JS version of `start()` returns a Promise, which will throw an exception if there are any exceptions thrown while evaluating the client provided selection sync query code block (not implemented on Android yet, but coming soon)

So, there are reasons for having callbacks, and reasons not to.  Given that adding this new API is a bit of a one-way door, I'd like to keep the callbacks, even if they don't report very meaningful errors, since we will likely need them later (probably for selective sync). 

If there are no objections to the above, this PR is ready to fly, and I'll merge as soon as I get an approval!

**Review by rjuliano — APPROVED** [review]
Ship it!!!🚀

File: aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java:176]: How about:
```suggestion
        Completable
            .create(subscriber -> {
                new Thread(() -> {
                   try {
                       orchestrator.start();
                       subscriber.onComplete();
                   } catch (DataStoreException failure) {
                       subscriber.onFailure(failure);
                   }
                }).start();
            })
            .blockingAwait(5, TimeUnit.SECONDS);
```
Or:
```suggestion
        Await
            .<VoidResult, DataStoreException>result((onResult, onError) -> {
                new Thread(() -> {
                   try {
                       orchestrator.start();
                       onResult.accept(VoidResult.instance());
                   } catch (DataStoreException failure) {
                       onError.accept(failure);
                   }
                }).start();
            });
```

**Review by jamesonwilliams — APPROVED** [review]

File: aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java
Comment (richardmcclellan) [reply aws-datastore/src/test/java/com/amplifyframework/datastore/syncengine/OrchestratorTest.java:176]: I like both of these better, thanks for the suggestions.  I'll go with the RxJava approach :).   RxJava feels cleaner than using a `CountDownLatch`.    I'd like to start using the [RxJava timeout operator](http://reactivex.io/documentation/operators/timeout.html) more too, instead of CountDownLatch, where possible.

**Pull request merged**
